### PR TITLE
Attach scroll/wheel to root: alternative scroll browser locking fix

### PR DESCRIPTION
**Pull request by nhunzaker**
This is an alternative solution to https://github.com/facebook/react/pull/9333 that attaches scroll/wheel listeners to the root container element.

So far, this prevents browser scroll locking in Chrome, Edge and Firefox. **It does not appear to fix locking on Safari**. However I think I might be getting inaccurate readings in BrowserStack and would really appreciate some help testing.

### Test Plan

1. Visit http://nh-react-scroll-root.surge.sh/scroll
2. Scroll around using the window for 10 seconds
3. Scroll around using the inner scrolling div for 10 seconds

**Expected:** No scroll locking

**Browsers**

- [x] Chrome 69
- [x] Edge 17
- [x] Firefox 61
- [ ] Safari 11.1

### CODE REVIEW

File: packages/react-dom/src/events/ReactBrowserEventEmitter.js [hunk 1]
```diff
@@ -124,25 +125,33 @@ function getListeningForDocument(mountAt: any) {
...
+        case TOP_WHEEL:
+          const rootListeners = getListenerTrackingFor(root);
+
+          if (!rootListeners.hasOwnProperty(dependency)) {
+            trapCapturedEvent(dependency, root);
+            rootListeners[dependency] = true;
+          }
+          return;
         case TOP_FOCUS:
         case TOP_BLUR:
```
Comment (nhunzaker) [inline packages/react-dom/src/events/ReactBrowserEventEmitter.js]: @gaearon you mentioned that touch events were essential (https://github.com/facebook/react/pull/9333#discussion_r212005901), would you like me to add them to the list here?

**Comment by gaearon** [comment]
I was thinking that if we do this, we should do it for all events. And behind a feature flag (it will be a pain to deploy at FB and would need to be breaking change).

File: packages/react-dom/src/events/ReactBrowserEventEmitter.js [hunk 2]
```diff
@@ -172,11 +181,11 @@ export function isListeningToAllDependencies(
   registrationName: string,
   mountAt: Document | Element,
 ) {
-  const isListening = getListeningForDocument(mountAt);
+  const isListening = getListenerTrackingFor(mountAt);
   const dependencies = registrationNameDependencies[registrationName];
   for (let i = 0; i < dependencies.length; i++) {
     const dependency = dependencies[i];
-    if (!(isListening.hasOwnProperty(dependency) && isListening[dependency])) {
```
Comment (nhunzaker) [inline packages/react-dom/src/events/ReactBrowserEventEmitter.js]: I'm going to revert this change. It isn't related to this work and should be done in a separate PR.

**Comment by react-linter** [comment]
ReactDOM: size: 0.0%, gzip: :small_red_triangle:+0.1%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: e106b8c44f9a81058f250c7ee37b61ae0094455e...9489e92a12036c9674abd76a71c38cbe79016b65</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | 0.0% | 0.0% | 649 KB | 649.23 KB | 152.36 KB | 152.41 KB | UMD_DEV
**react-dom.production.min.js** | **0.0%** | **:small_red_triangle:+0.1%** | **95.3 KB** | **95.33 KB** | **31.17 KB** | **31.21 KB** | **UMD_PROD**
react-dom.development.js | 0.0% | 0.0% | 645.14 KB | 645.37 KB | 151.24 KB | 151.3 KB | NODE_DEV
**react-dom.production.min.js** | **0.0%** | **0.0%** | **95.29 KB** | **95.32 KB** | **30.89 KB** | **30.91 KB** | **NODE_PROD**
ReactDOM-dev.js | 0.0% | 0.0% | 652.45 KB | 652.68 KB | 149.39 KB | 149.46 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.2%** | **287.9 KB** | **288.21 KB** | **53.34 KB** | **53.43 KB** | **FB_WWW_PROD**
react-dom.profiling.min.js | 0.0% | 0.0% | 96.49 KB | 96.52 KB | 31.29 KB | 31.3 KB | NODE_PROFILING
ReactDOM-profiling.js | +0.1% | +0.2% | 290.25 KB | 290.55 KB | 53.94 KB | 54.03 KB | FB_WWW_PROFILING

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by nhunzaker** [comment]
Cool. Doing that exposes a few failures. Particularly around mouse enter and leave because it needs to pick up changes outside of React. That might be fixable by attaching to the owner document only for those events.

**Comment by gaearon** [comment]
Note past attempt in https://github.com/facebook/react/pull/8117 and the uncovered difficulties.

**Comment by nhunzaker** [comment]
Yeah. This is going to take some planning. Maybe we can talk about it at the DOM team meeting.

**Comment by nhunzaker** [comment]
Closing this out in favor of #9333. It looks like attaching to the root prevents the outer scroll region (the window in the fixture case) from being optimized in Edge and Safari. In #9333, I think I've figured out how to safely attach local listeners.
